cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ZookeeperScaler.java

```diff
@@ -0,0 +1,351 @@
+            return Future.succeededFuture();
+        }
+    }
+
+    /**
+     * Gets the current configuration from Zookeeper.
+     *
+     * @return  Future containing Map with the current Zookeeper configuration
+     */
+    private Future<Map<String, String>> getConfig()    {
+        Promise<Map<String, String>> configPromise = Promise.promise();
+
+        vertx.executeBlocking(promise -> {
+            try {
+                byte[] config = zkAdmin.getConfig(false, null);
+                Map<String, String> servers = parseConfig(config);
+                log.debug("Current Zookeeper configuration is {}", servers);
+                promise.complete(servers);
+            } catch (KeeperException | InterruptedException e)    {
+                log.warn("Failed to get current Zookeeper server configuration", e);
+                promise.fail(new RuntimeException("Failed to get current Zookeeper server configuration", e));
+            }
+        }, false, configPromise);
+
+        return configPromise.future();
+    }
+
+    /**
+     * Updates the configuration in the Zookeeper cluster
+     *
+     * @param newServers    New configuration which will be used for the update
+     * @return              Future with the updated configuration
+     */
+    private Future<Map<String, String>> updateConfig(Map<String, String> newServers)    {
+        Promise<Map<String, String>> configPromise = Promise.promise();
+
+        vertx.executeBlocking(promise -> {
+            try {
+                log.debug("Updating Zookeeper configuration to {}", newServers);
+                byte[] newConfig = zkAdmin.reconfigure(null, null, serversMapToList(newServers), -1, null);
+                Map<String, String> servers = parseConfig(newConfig);
+
+                log.debug("New Zookeeper configuration is {}", servers);
+                promise.complete(servers);
+            } catch (KeeperException | InterruptedException e)    {
+                log.warn("Failed to update Zookeeper server configuration", e);
+                promise.fail(new RuntimeException("Failed to update Zookeeper server configuration", e));
+            }
+        }, false, configPromise);
+
+        return configPromise.future();
+    }
+
+    /**
+     * Closes the Zookeeper connection asynchronously
+     *
+     * @return
+     */
+    private Future<Void> closeConnectionAsync() {
+        Promise<Void> closePromise = Promise.promise();
+
+        log.debug("Closing Zookeeper connection to {}", zookeeperConnectionString);
+
+        vertx.executeBlocking(promise -> {
+            closeConnectionSync();
+            promise.complete();
+        }, true, closePromise);
+
+        return closePromise.future();
+    }
+
+    /**
+     * Closes the Zookeeper connection synchronously
+     */
+    private void closeConnectionSync() {
+        if (zkAdmin != null)    {
+            try {
+                zkAdmin.close();
```

Comment:
Timeout?

Reply:
The `close` call is immadiate without any timeout. It has a variant with timeout which I found not to be working when any problems happen to the connection so I decided to use this option.